Feature: Change document thumbnails to WebP

stumpylog opened the pull request:
## Proposed change

As discussed in #1125, this PR updates the generation of thumbnails to output in WebP format, instead of PNG.  Existing PNG thumbnails will remain untouched, with a system check info suggesting, but not requiring conversion to WebP.  WebP can be a large space savings for the storage and faster loading for the Documents view.  There isn't really a good reason I can image to use a lossless format for thumbnails.

For the conversion, a new management command is created, which converts existing PNG to WebP.

Note that due to the thumbnail caching, a cache clear would be needed to get the converted WebP thumbnails, otherwise the cached thumbnails are used.  If someone with HTTP/django knowledge wants to sort out the TODO regarding ETags, this would be a good reason.

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (please explain)

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#code-formatting-with-pre-commit-hooks).
- [ ] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

shamoon committed:
webp thumbnail support with png fallback

stumpylog committed:
Minor tweaks to getting the document thumbnail path.  Adds text thumbnail as webp

stumpylog committed:
Adds untested manual conversion command

stumpylog committed:
Corrects the logic of thumbnail path to account for both getting existing path or building expected path

stumpylog committed:
Corrects the functionality of the webp conversion script

stumpylog committed:
Adds information system check for PNG thumbnail existence

stumpylog committed:
Minor tweak to determining content type

stumpylog committed:
Fixes existing testing, adds test coverage of new command

shamoon commented:
This works great!
```
INFOS:
?: PNG thumbnails found, consider running convert_thumbnails to convert to WebP
```

```
Conversion to WebP completed
Conversion completed in 3.942s
```

I think this is the best solution and we can document the change in the release so people can know to run the command. Its perfect "progressive enhancement" IMHO. Even for users who never run the command from now on thumbnails will be webp which should be beneficial. Overall I'm really happy with this!

stumpylog committed:
Adds quick documentation of new functionality

shamoon commented:
Does this break the `document_thumbnails` command which is supposed to re-generate thumbnails, presumably yes because if its a webp it doesnt re-generate now (`get_optimised_thumbnail` ignores non-pngs). Unless Im confused? `document_archiver` users that too, but Im not sure why the archiver would re-run the optimization anyway so maybe that ones ok?

stumpylog committed:
Fixes the re-generation of thumbnails making webp file, but named as PNG files

stumpylog commented:
Yeah, it'll be a little weird.  It will make a WebP file, but then save it as `.png` because the PNG file exists already.  So the thumbnail is removed first now, to fix the filename

adam-ah commented on src/documents/management/commands/document_thumbnails.py:
```diff
@@ -21,6 +22,13 @@ def _process_document(doc_in):
+        existing_thumbnail = Path(document.thumbnail_path).resolve()
+
+        # Remove an existing PNG format thumbnail, if it existed
+        if existing_thumbnail.exists() and existing_thumbnail.suffix == ".png":
```
this code is a bit surprising here - should it go to `get_optimised_thumbnail` instead?

adam-ah commented on src/documents/models.py:
```diff
@@ -292,12 +293,33 @@ def file_type(self):
+        webp_file_path = os.path.join(settings.THUMBNAIL_DIR, webp_file_name)
+        png_file_path = os.path.join(settings.THUMBNAIL_DIR, png_file_name)
+
+        # 1. Assume the thumbnail is WebP
```
I think this whole block (310-321) can be replaced by for less file checks, readability, less nested `if`s, and less `not if` logics:
```
thumb = webp_file_path
if os.path.exists(png_file_path):
    thumb = png_file_path
```

it seems the `if not os.path.exists(webp_file_path):` check is irrelevant.

We *only* set the PNG thumb path if that file exists, no need to check if webp exists

adam-ah commented on src/documents/views.py:
```diff
@@ -362,7 +362,13 @@ def thumb(self, request, pk=None):
             #  if available
-            return HttpResponse(handle, content_type="image/png")
+            thumbnail_path = doc.thumbnail_path
+            if os.path.splitext(thumbnail_path)[1] == ".webp":
```
it seems `thumbnail_path` is only used once in the next line, consider not creating the variable and just inlining `doc.thumbnail_path` in here

adam-ah reviewed (commented):
This is great work, I left a few comments for you to consider (take/leave)

adam-ah commented:
> I think this is the best solution and we can document the change

I'm a little bit unsure about this; what is the chance people would even know about the new webp feature, based on checking the logs or reading the documentation?

Given the assumptions previously about the typical user and typical usecase, most existing users will not benefit from these changes (won't read the docs, won't check the logs, and won't know how to run the migration scripts).

> Conversion completed in 3.942s

@shamoon Are we assuming *most* users will have extremely longer conversion times so it's worth logging about it rather than actually doing the conversion? 

I fear we are adding these changes to cater for the extreme scenario (slow server, 1000s of documents, huge PNG files, ) rather than the typical user with the typical usecase (normal server, avg few 100 documents, not all PNGs are huge)

shamoon commented:
I don't think there's a definitive answer. There are pros/cons either way.

As always would be valuable to hear others' thoughts...

stumpylog commented on src/documents/checks.py:
```diff
@@ -67,3 +68,23 @@ def parser_check(app_configs, **kwargs):
+                        "to convert to WebP",
+                    ),
+                ]
+        return []
```
That's [required by django](https://docs.djangoproject.com/en/4.0/topics/checks/#messages)

stumpylog commented on src/documents/management/commands/convert_thumbnails.py:
```diff
@@ -0,0 +1,94 @@
+
+                            self.stdout.write(
+                                self.style.SUCCESS(
+                                    "Conversion to WebP completed",
```
`converted_thumbnail` in this case will be a mostly random temporary directory, and the whole directory will be removed after the command completes.  So that doesn't seem relevant or useful to a user as they can't access the file after completion.

adam-ah replied:
These logs are not useful for the users per se: they will never want to look at or understand these.

However, these logs are really useful for bug reports when something goes wrong - and arguably having more information that might help to find problems is better than a static message.

shamoon replied:
When you view the lines before its clear which youre talking about, seems OK to me as-is

shamoon commented on src/documents/checks.py:
```diff
@@ -67,3 +68,23 @@ def parser_check(app_configs, **kwargs):
+
+
+@register()
+def png_thumbnail_check(app_configs, **kwargs):
```
I’m not sure how expensive this check is, completes in ms and there are a bunch of checks that run on startup so I think this is reasonable

stumpylog commented:
With 496 documents, the conversion took ~36s for me.  With a multiprocessing Pool working on the conversions, it drops to ~10s.  That wouldn't be terrible as a one-time migration, I think even at larger file counts or slower hardware.

An advantage of one time, it does let the code simplify out more, no need to check `.png` everywhere, it's `.webp` all places.

stumpylog committed:
Entirely removes the optipng, updates ghostscript fall back to also use WebP.  Updates the conversion to use a multiprocessing pool

stumpylog committed:
Converts the conversion into a database migration

stumpylog committed:
Removes last vestiges of PNG from the tests, code, docs and samples

stumpylog commented:
Ok, I've changed my mind and the conversion is now done as a one time migration at startup.  I found a couple existing migrations such as 0012_auto_20160305_0040 (which generated thumbnails for all documents) that helped convince me the occasional slow migration is Ok.  To help make it less slow, a process pool is used to work multiple files in parallel.

So with that:
- No more PNG vs WebP handling in the code, everything is WebP
- Fixed up all the tests and sample files which were expecting or relying on PNG
- Removed thumbnail optimization and related binaries
- Ghostscript thumbnail fallback also should produce WebP thumbnails

The only thing outstanding I can think of is adding ETag to the thumbnail view, so the cache will be invalidated for the old thumbnails.  But since the images are the same, it's not a big deal.

adam-ah commented on src/documents/migrations/1021_webp_thumbnail_conversion.py:
```diff
@@ -0,0 +1,107 @@
+                (existing_thumbnail, converted_thumbnail),
+            )
+
+        if len(work_packages):
```
Consider using the [PEP8 recommendation](https://peps.python.org/pep-0008/#programming-recommendations)
`if work_packages:`

stumpylog replied:
I'm going to keep with the Zen on this one.  I hate guessing what the if is actually checking.

adam-ah commented on src/documents/views.py:
```diff
@@ -362,7 +362,9 @@ def thumb(self, request, pk=None):
-            return HttpResponse(handle, content_type="image/png")
+            content_type = "image/webp"
+
+            return HttpResponse(handle, content_type=content_type)
```
perhaps just inline it as the `content_type` variable is not used anywhere else, so:

`return HttpResponse(handle, content_type="image/webp")`

stumpylog replied:
Right, one last bit I missed there.

adam-ah reviewed (commented):
This looks great, much appreciated. So much easier to follow now that everything is webp. Only two optional stylistic comments

shamoon commented:
Great work as usual stumpy! The optimizations to speed it up as a migration are awesome

shamoon reviewed (approved):
Pending comments seem reasonable too but I’ll leave that up to you

stumpylog committed:
Removes one last portion of PNG vs WebP

shamoon commented:
Bravo!
